Impl tower::Service for tonic::transport::Channel

## Feature Request

### Motivation

A type that implements `tower::Service` in the proper way implements `GrpcService`, but not the other way around.

[`Channel`](https://docs.rs/tonic/0.3.1/tonic/transport/channel/struct.Channel.html) is a special implementation of `GrpcService`. It can potentially implement `Service` as well (making such a wrapper type is easy), but currently it doesn't. Due to this, we can't make use of various `tower`'s wrapper services for a channel.

```rust
#[tokio::main]
async fn main() -> Result<(), Box<dyn std::error::Error>> {
    let conn = tonic::transport::Endpoint::new("http://[::1]:50051")?
        .connect()
        .await?;
    // Wrap a channel with Timeout
    let conn = Timeout::new(conn, Duration::from_millis(1000));
    let mut client = GreeterClient::new(conn);  // Error!
    let request = tonic::Request::new(HelloRequest {
        name: "Tonic".into(),
    });

    let response = client.say_hello(request).await?;

    println!("RESPONSE={:?}", response);

    Ok(())
}
```

```
error[E0277]: the trait bound `tonic::transport::Channel: tower::Service<http::Request<tonic::body::BoxBody>>` is not satisfied
  --> src/client.rs:40:41
   |
40 |     let mut client = GreeterClient::new(conn);
   |                                         ^^^^ the trait `tower::Service<http::Request<tonic::body::BoxBody>>` is not implemented for `tonic::transport::Channel`
   |
   = note: required because of the requirements on the impl of `tower::Service<http::Request<tonic::body::BoxBody>>` for `tower::tower_timeout::Timeout<tonic::transport::Channel>`
   = note: required because of the requirements on the impl of `tonic::client::GrpcService<tonic::body::BoxBody>` for `tower::tower_timeout::Timeout<tonic::transport::Channel>`

```


### Proposal

Let's implement `tower::Service<http::Request<BoxBody>>` for `Channel`, so that we can make use of `tower` service wrappers more easily.

### Alternatives

<!--
Are there other ways to solve this problem that you've considered? What are
their potential drawbacks? Why was the proposed solution chosen over these
alternatives?
-->

To achieve the same goal, users may write a small wrapper for `Channel` which implements `Service`.

```rust
struct ChannelWrapper(Channel);

impl Service<http::Request<BoxBody>> for ChannelWrapper {
    type Response = http::Response<Body>;
    type Error = tonic::transport::Error;
    type Future = tonic::transport::channel::ResponseFuture;

    fn poll_ready(&mut self, cx: &mut Context<'_>) -> Poll<Result<(), Self::Error>> {
        self.0.poll_ready(cx)
    }

    fn call(&mut self, req: http::Request<BoxBody>) -> Self::Future {
        let mut channel = self.0.clone();

        channel.call(req)
    }
}
```

```rust
#[tokio::main]
async fn main() -> Result<(), Box<dyn std::error::Error>> {
    let conn = tonic::transport::Endpoint::new("http://[::1]:50051")?
        .connect()
        .await?;
    let conn = ChannelWrapper(conn);
    let conn = Timeout::new(conn, Duration::from_millis(1000));
    let mut client = GreeterClient::new(conn);
...
```

We can find [a similar one in `examples`](https://github.com/hyperium/tonic/blob/master/examples/src/tower/client.rs). But providing a canonical `Service` implementation for `Channel` would simplify user code?

--- a/examples/Cargo.toml
+++ b/examples/Cargo.toml
@@ -75,6 +75,14 @@ name = "tower-client"
 path = "src/tower/client.rs"
 
 [[bin]]
+name = "timeout-server"
+path = "src/timeout/server.rs"
+
+[[bin]]
+name = "timeout-client"
+path = "src/timeout/client.rs"
+
+[[bin]]
 name = "multiplex-server"
 path = "src/multiplex/server.rs"
 

--- a/examples/src/timeout/client.rs
+++ b/examples/src/timeout/client.rs
@@ -1,0 +1,28 @@
+use hello_world::greeter_client::GreeterClient;
+use hello_world::HelloRequest;
+use std::time::Duration;
+use tower::timeout::Timeout;
+
+use tonic::transport::Channel;
+
+pub mod hello_world {
+    tonic::include_proto!("helloworld");
+}
+
+#[tokio::main]
+async fn main() -> Result<(), Box<dyn std::error::Error>> {
+    let channel = Channel::from_static("http://[::1]:50051").connect().await?;
+    let timeout_channel = Timeout::new(channel, Duration::from_millis(1000));
+
+    let mut client = GreeterClient::new(timeout_channel);
+
+    let request = tonic::Request::new(HelloRequest {
+        name: "Tonic".into(),
+    });
+
+    let response = client.say_hello(request).await?;
+
+    println!("RESPONSE={:?}", response);
+
+    Ok(())
+}

--- a/examples/src/timeout/server.rs
+++ b/examples/src/timeout/server.rs
@@ -1,0 +1,45 @@
+use std::time::Duration;
+use tokio::time::delay_for;
+use tonic::{transport::Server, Request, Response, Status};
+
+use hello_world::greeter_server::{Greeter, GreeterServer};
+use hello_world::{HelloReply, HelloRequest};
+
+pub mod hello_world {
+    tonic::include_proto!("helloworld");
+}
+
+#[derive(Default)]
+pub struct MyGreeter {}
+
+#[tonic::async_trait]
+impl Greeter for MyGreeter {
+    async fn say_hello(
+        &self,
+        request: Request<HelloRequest>,
+    ) -> Result<Response<HelloReply>, Status> {
+        println!("Got a request from {:?}", request.remote_addr());
+
+        delay_for(Duration::from_millis(5000)).await;
+
+        let reply = hello_world::HelloReply {
+            message: format!("Hello {}!", request.into_inner().name),
+        };
+        Ok(Response::new(reply))
+    }
+}
+
+#[tokio::main]
+async fn main() -> Result<(), Box<dyn std::error::Error>> {
+    let addr = "[::1]:50051".parse().unwrap();
+    let greeter = MyGreeter::default();
+
+    println!("GreeterServer listening on {}", addr);
+
+    Server::builder()
+        .add_service(GreeterServer::new(greeter))
+        .serve(addr)
+        .await?;
+
+    Ok(())
+}

--- a/examples/src/tower/client.rs
+++ b/examples/src/tower/client.rs
@@ -32,7 +32,6 @@ mod service {
     use std::pin::Pin;
     use std::task::{Context, Poll};
     use tonic::body::BoxBody;
-    use tonic::client::GrpcService;
     use tonic::transport::Body;
     use tonic::transport::Channel;
     use tower::Service;

--- a/tonic/src/transport/channel/mod.rs
+++ b/tonic/src/transport/channel/mod.rs
@@ -10,7 +10,7 @@ pub use endpoint::Endpoint;
 pub use tls::ClientTlsConfig;
 
 use super::service::{Connection, DynamicServiceStream};
-use crate::{body::BoxBody, client::GrpcService};
+use crate::body::BoxBody;
 use bytes::Bytes;
 use http::{
     uri::{InvalidUri, Uri},
@@ -177,17 +177,18 @@ impl Channel {
     }
 }
 
-impl GrpcService<BoxBody> for Channel {
-    type ResponseBody = hyper::Body;
+impl Service<http::Request<BoxBody>> for Channel {
+    type Response = http::Response<super::Body>;
     type Error = super::Error;
     type Future = ResponseFuture;
 
     fn poll_ready(&mut self, cx: &mut Context<'_>) -> Poll<Result<(), Self::Error>> {
-        GrpcService::poll_ready(&mut self.svc, cx).map_err(super::Error::from_source)
+        Service::poll_ready(&mut self.svc, cx).map_err(super::Error::from_source)
     }
 
-    fn call(&mut self, request: Request<BoxBody>) -> Self::Future {
-        let inner = GrpcService::call(&mut self.svc, request);
+    fn call(&mut self, request: http::Request<BoxBody>) -> Self::Future {
+        let inner = Service::call(&mut self.svc, request);
+
         ResponseFuture { inner }
     }
 }

